// @sendbird/uikit-message-template/src/core/renderer.tsx
import React from 'react';

import type { BasicProps, ComponentType, ComponentsUnion, GetProperties } from '../types/components';

type FuncComponent<Props, ParsedStyle> = (props: BasicProps<Props, ParsedStyle>) => React.ReactElement | null;

type ComponentProperties<Type extends ComponentType> = GetProperties<Type, ComponentsUnion>;

interface CreateRendererParams<ParsedStyle> {
  views?: {
    [component in ComponentType]?: FuncComponent<ComponentProperties<component>, ParsedStyle>;
  };
}

export type Renderer<ParsedStyle> = {
  [component in ComponentType]: FuncComponent<ComponentProperties<component>, ParsedStyle>;
};

const FRAGMENT = ({ children }: React.PropsWithChildren<unknown>) => <React.Fragment>{children}</React.Fragment>;

export function createRenderer<ParsedStyle = object | string>(
  params?: CreateRendererParams<ParsedStyle>,
): Renderer<ParsedStyle> {
  return {
    box: params?.views?.box || FRAGMENT,
    text: params?.views?.text || FRAGMENT,
    image: params?.views?.image || FRAGMENT,
    imageButton: params?.views?.imageButton || FRAGMENT,
    textButton: params?.views?.textButton || FRAGMENT,
    carouselView: params?.views?.carouselView || FRAGMENT,
  };
}
