// @sendbird/react-uikit-message-template-view/src/util/index.ts
import { ComponentType, FlexSizeSpecValue, SizeSpec, View } from '@sendbird/uikit-message-template';

export function isWrappedText(view: View) {
  return view.type === ComponentType.Text && view.width?.value == FlexSizeSpecValue.WrapContent;
}

// Compares the size spec to the option and returns whether they are equal or not.
export const isSizeCompatibleWithOption = ({
  size,
  option,
}: {
  size: SizeSpec;
  option: 'wrap' | 'fill' | 'fixed';
}): boolean => {
  if (option === 'wrap') {
    return size.type === 'flex' && size.value == FlexSizeSpecValue.WrapContent;
  }

  if (option === 'fill') {
    return size.type === 'flex' && size.value == FlexSizeSpecValue.FillParent;
  }

  return size.type === 'fixed';
};

export const isGif = (url: string | undefined) => {
  if (typeof url !== 'string') return false;
  const extension = url.split('.').pop()?.toLowerCase();
  return extension === 'gif';
};

export const isNumber = (val?: string | number | null): val is number => {
  if (typeof val === 'string') {
    return !Number.isNaN(Number(val));
  }
  return typeof val === 'number' && !Number.isNaN(val);
};
